### DISCUSSION TITLE
Completely disable osg culling

### DISCUSSION THREAD

**Discussion by rickyviking** [openscenegraph/OpenSceneGraph#1193] [Q&A]
Hi all,

For a particular use case I need to completely disable all culling performed by OpenSceneGraph. 
I thought it was enough (for a single camera viewer application) to call:

```_camera->setCullingMode(osg::CullSettings::NO_CULLING);```

but some drawables are still culled away. 

To confirm this behavior I have modified osgViewer adding an event handler to modify the CullingMode at runtime, and when the CullingMode is set to NO_CULLING, the rendered drawable and polygon count increase as expected, but are still lower than the full counters shown by the VIEW stats.
See the attached screenshots which show the stats using NO_CULLING and DEFAULT_CULLING settings respectively for a sample model.

Which is the right way to completely disable all culling (and thus see the same poly count in the camera and view stats)?

Thank you,
Riccardo

DEFAULT_CULLING
![osg_viewer_DEFAULT_CULLING](https://user-images.githubusercontent.com/2272034/211936461-8ad58dee-a008-4e39-a83a-57025bf492f3.jpg)  

NO_CULLING
![osg_viewer_NO_CULLING](https://user-images.githubusercontent.com/2272034/211936464-7b601423-d9ae-4599-a449-a41728b42b3b.jpg)

**Comment by robertosfield** [comment]
Could it be that you have Switch/Sequence statements in the scene?

**Reply by rickyviking** [reply]
No, it is a very optimized geometry-only graph.
It includes only Group and (fast) Geometry nodes, no transforms, no display lists.

**Reply by robertosfield** [reply]
It may well be just down to the way the view stats vs cull stats are collected as they have to be done with different mechanisms.  If you really feel you need to know then just add some debugging into the OSG and count the data directly yourself.

Or... just realize that a fps of over 60 is more than enough so pointless trying to optimize further, and if performance really is critical then you should be using Vulkan/VSG not the OpenGL/OSG.

**Reply by rickyviking** [reply]
OK @robertosfield I'll try to verify if the problem is due only to the stats computation... 
so can you confirm that switching the camera CullSettings to NO_CULLING should be enough?

In this particular case I'm not trying to optimize further, but I just need to make sure that the whole scene is rendered without any culling performed by OpenSceneGraph. 

Thanks

**Reply by robertosfield** [reply]
I can't confirm anything without diving into the code and doing on my debug testing.  I have other work to do so I'll not waste my time something that doesn't provide any useful benefit to the wider community.  You have the source code so if you want to dive in feel free.

**Reply by rickyviking** [reply]
A quick update with my findings in case someone else might be interested.

The stats are ok, some culling actually still happens when the NO_CULLING setting is set on the camera.
This is due to the Near/Far planes computation routine: if a Drawable is found to be behind the eye point, not only it does not contribute to the planes calculation, but it is also culled away. 
See:
- https://github.com/openscenegraph/OpenSceneGraph/blob/master/src/osgUtil/CullVisitor.cpp#L927
- https://github.com/openscenegraph/OpenSceneGraph/blob/master/src/osgUtil/CullVisitor.cpp#L754  
  
After disabling the near/far computation as well (using osg::CullSettings::DO_NOT_COMPUTE_NEAR_FAR) the Camera and View stats correspond exactly, and I've counter checked that the geometries are actually sent into the render pipeline.
